yo1k/sandbox/exercises/CSC_algorithms/sorting.py

```diff
@@ -1,131 +1,158 @@
+        assert l_idx <= end_subseq
+        seq[l_idx] = tmp_seq[i]
         l_idx += 1
-    while r_idx < r_end:
-        merged_array.append(array_right[r_idx])
-        r_idx += 1
-    return merged_array
 
 
-def merge_sort(array: MutableSequence[Any]) -> MutableSequence[Any]:
+def merge_sorted_seqs(
+        seq: MutableSequence[Any],
+        tmp_seq: MutableSequence[Any],
+        l_begin_seq: int,
+        l_end_seq: int,
+        r_begin_seq: int,
+        r_end_seq: int) -> None:
+    """Merges two sorted seqs in ascending order."""
+    l_seq_idx: int = l_begin_seq
+    r_seq_idx: int = r_begin_seq
+    l_seq_len: int = l_end_seq - l_begin_seq + 1
+    r_seq_len: int = r_end_seq - r_begin_seq + 1
+    tmp_seq_idx = 0
+
+    while l_seq_idx - l_begin_seq < l_seq_len and r_seq_idx - r_begin_seq < r_seq_len:
+        if seq[l_seq_idx] <= seq[r_seq_idx]:
+            tmp_seq[tmp_seq_idx] = seq[l_seq_idx]
+            tmp_seq_idx += 1
+            l_seq_idx += 1
+        else:
+            tmp_seq[tmp_seq_idx] = seq[r_seq_idx]
+            tmp_seq_idx += 1
+            r_seq_idx += 1
+    while l_seq_idx - l_begin_seq < l_seq_len:
+        tmp_seq[tmp_seq_idx] = seq[l_seq_idx]
+        tmp_seq_idx += 1
+        l_seq_idx += 1
+    while r_seq_idx - r_begin_seq < r_seq_len:
+        tmp_seq[tmp_seq_idx] = seq[r_seq_idx]
+        tmp_seq_idx += 1
+        r_seq_idx += 1
+
+    replace_subseq(seq, l_begin_seq, r_end_seq, tmp_seq)
+
+
+def merge_sort(seq: MutableSequence[Any]) -> MutableSequence[Any]:
     """Mergesort.
 
-    This algorithm has complexity O(n*log(n)) on condition
-    when we do not take into account overheads of creating new arrays in the process of sorting,
-    only comparison operation."""
-    size: int = len(array)
-    l_size: int = size // 2
-    r_size: int = size // 2
-    if size % 2 != 0:
-        l_size += 1
-
-    if r_size == 0:
-        return array
-    l_array = array[: l_size]
-    r_array = array[l_size:]
-    assert len(l_array) == l_size and len(r_array) == r_size
-    l_array = merge_sort(l_array)
-    r_array = merge_sort(r_array)
-    sorted_array = merge_sorted_arrays(l_array, r_array)
-    return sorted_array
-
-
-def partition(array: MutableSequence[Any], l_idx: int, r_idx: int) -> int:
-    """Changes `array` to: elements <= pivot, then a pivot element, elements > pivot.
-
-    Without loss of generality the pivot is taken as the `l_idx`-th element of the `array`.
+    This algorithm has complexity O(n*log(n)).
+    """
+    size = len(seq)
+    tmp_seq = [None for _ in range(size)]
+    if size == 0:
+        return seq
+
+    def _merge_sort(_seq: MutableSequence[Any], begin: int, end: int) -> MutableSequence[Any]:
+        delimetr_idx = (end - begin) // 2
+        l_seq_begin = begin
+        l_seq_end = l_seq_begin + delimetr_idx
+        r_seq_begin = l_seq_end + 1
+        r_seq_end = end
+
+        if r_seq_end - l_seq_begin == 0:
+            return _seq
```

Comment:
It appears the code I [commented](https://github.com/Yo1k/sandbox-python/pull/3#discussion_r850155336) previously was changed to this code. And this condition is still true only if the size of the subsequence is 1, i.e., `end - begin == 1`, which we should check at the very beginning of the `_merge_sort` method instead of having this code here.

Also, the condition `if size == 0` at the beginning of the `merge_sort` method should be replaced with `if size == 0 or size == 1`.

Reply:
Done.

> And this condition is still true only if the size of the subsequence is 1, i.e., end - begin == 1

Only end - begin == 0, because both of these indices are included in the `subseq`.